pallets/collator-selection/src/lib.rs

```diff
@@ -375,7 +405,16 @@ pub mod pallet {
 				index,
 				<frame_system::Pallet<T>>::block_number(),
 			);
-			Some(Self::assemble_collators())
+
+			let candidates_len_before = Self::candidates().len();
+			let result = Self::assemble_collators();
```

Comment:
This is not a required change but I want to point out that assemble_collators is having side effects which makes the code here a bit hard to follow. 

reading the candidates once, and getting passing it through two functions where we:
1. kick stale ones
2. prefix with invulnerables

```
let mut candidates = Self::candidates();
Self::kick_stale_candidates(&mut candidates);
let collators = Self:: assemble_collators(candidates);
```

is somewhat cleaner. WDYT?

Reply:
ya I think so, actually if I return just a vec of accounts from kick stale then I can get rid of the map in assemble collators too